### PR TITLE
cmd/contour: Replace check() by  kingpin.Fatalf() in cli.go

### LINKED ISSUE #2811: Replace check with Fatal logs
**Please describe the problem you have**

There are quite a few places in Contour that call a local `check` helper that prints to standard error and exits. This approach loses error context and makes it difficult for aggregated log collectors to recognize that this is a fatal error message. We should replace these with fatal level logs.

### PR DESCRIPTION
**Pull request by ShaileshSurya**
There are quite a few places in Contour that call a
local check helper that prints to standard error and exits.
This approach loses error context and makes it difficult for
aggregated log collectors to recognize that this is a fatal error message

This commit replaces check functions in cmd/contour/cli.go by log.Fatal

Fixes #2811

Signed-off-by: Shailesh Suryawanshi <ss.shailesh28@gmail.com>

### CODE REVIEW

**Review by jpeach — CHANGES_REQUESTED** [review]
This code is only used by the `contour cli *` subcommands, so exiting manually is appropriate here. The `log.Fatal` in `(*Client) dial()` is actually a bug.

We have a few options with this code:
1. Assume that it is library code and propagate errors out to the caller, who can exit with an error
1. Replace `check` with a function `ErrExit` which does the same thing, but has a clearer name
1. Replace `check` with calls to [kingpin.FatalIfError](https://godoc.org/github.com/alecthomas/kingpin#FatalIfError)
1. Leave it for now :)

The first option is a fair bit of churn, so you may want to consider the other options :)

**Comment by ShaileshSurya** [comment]
Hi @jpeach, 

I find the third option `Replace check with calls to the kingpin.FatalIfError` intuitive. I will make the changes accordingly and update the PR. 

I will also update the existing `log.fatal` with the kingpin.FatalIfError. 

Thank you for the suggestions.

File: cmd/contour/cli.go [hunk 1]
```diff
@@ -48,20 +49,16 @@ func (c *Client) dial() *grpc.ClientConn {
...
-		if err != nil {
-			log.Fatalf("err: %v", err)
-		}
+		kingpin.FatalIfError(err, "err: %v", err)
 
 		// Append the certificates from the CA
 		if ok := certPool.AppendCertsFromPEM(ca); !ok {
 			// TODO(nyoung) OMG yuck, thanks for this, crypto/tls. Suggestions on alternates welcomed.
-			log.Fatal("failed to append ca certs")
+			kingpin.Fatalf("failed to append ca certs")
```
Comment (ShaileshSurya) [inline cmd/contour/cli.go:61]: I am not very sure about this. I checked kingpins' documentation and could not find `kingpin.Fatal`. Please advise.

File: go.sum [hunk 2]
```diff
@@ -24,6 +24,9 @@ github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578/go.mod h1:uGdko
 github.com/agnivade/levenshtein v1.0.1/go.mod h1:CURSv5d9Uaml+FovSIICkLbAUZ9S4RqaHDIsdSBg7lM=
 github.com/ahmetb/gen-crd-api-reference-docs v0.1.5 h1:OU+AFpBEhyclrQGx4I6zpCx5WvXiKqvFeeOASOmhKCY=
 github.com/ahmetb/gen-crd-api-reference-docs v0.1.5/go.mod h1:P/XzJ+c2+khJKNKABcm2biRwk2QAuwbLf8DlXuaL7WM=
+github.com/alecthomas/kingpin v1.3.7 h1:GLMgiQ5nZb3rg5pozvF7KDbseQ12eNO1p7bB2t5K1fw=
+github.com/alecthomas/kingpin v2.2.6+incompatible h1:5svnBTFgJjZvGKyYBtMB0+m5wvrbUHiqye8wRJMlnYI=
```
Comment (ShaileshSurya) [inline go.sum:28]: I see this was not needed as this library is already imported as `kingpin "gopkg.in/alecthomas/kingpin.v2"`

https://github.com/projectcontour/contour/blob/0d684bc09c10668fc7d139dc5e25919d51498b1f/cmd/contour/contour.go#L24

Will make the changes when other code is reviewed.

File: go.mod [hunk 3]
```diff
@@ -4,6 +4,7 @@ go 1.14
 
 require (
 	github.com/ahmetb/gen-crd-api-reference-docs v0.1.5
+	github.com/alecthomas/kingpin v2.2.6+incompatible
```
Comment (ShaileshSurya) [inline go.mod:7]: I see this was not needed as this library is already imported as `kingpin "gopkg.in/alecthomas/kingpin.v2"`

https://github.com/projectcontour/contour/blob/0d684bc09c10668fc7d139dc5e25919d51498b1f/cmd/contour/contour.go#L24

Will make the changes once other code is reviewed.

**Review by jpeach — CHANGES_REQUESTED** [review]
Good progress, just need to update the exit messages. Can you manually trigger some of then and show the output?

Can you please update the `log.Fatal` on line 48 as well?

File: cmd/contour/cli.go [hunk 4]
```diff
@@ -126,10 +124,10 @@ func watchstream(st stream, typeURL string, resources []string) {
 			ResourceNames: resources,
 		}
 		err := st.Send(req)
-		check(err)
+		kingpin.FatalIfError(err, "")
```
Comment (ShaileshSurya) [inline cmd/contour/cli.go:127]: Please Check https://play.golang.org/p/1WiDkfYYoLp for the output.

File: cmd/contour/cli.go [hunk 5]
```diff
@@ -45,21 +44,20 @@ func (c *Client) dial() *grpc.ClientConn {
...
 		// If one of the three TLS commands is not empty, they all must be not empty
 		if !(c.CAFile != "" && c.ClientCert != "" && c.ClientKey != "") {
-			log.Fatal("You must supply all three TLS parameters - --cafile, --cert-file, --key-file, or none of them.")
+			kingpin.Fatalf("You must supply all three TLS parameters - --cafile, --cert-file, --key-file, or none of them.")
 		}
 		// Load the client certificates from disk
 		certificate, err := tls.LoadX509KeyPair(c.ClientCert, c.ClientKey)
-		check(err)
-
+		kingpin.FatalIfError(err, "")
```
Comment (ShaileshSurya) [inline cmd/contour/cli.go:51]: Please check https://play.golang.org/p/1WiDkfYYoLp for the output.

File: go.sum (hunk 2 above)
Comment (ShaileshSurya) [reply go.sum]: Removed This.

File: go.mod (hunk 3 above)
Comment (ShaileshSurya) [reply go.mod]: Done.

File: cmd/contour/cli.go [hunk 6]
```diff
@@ -17,11 +17,11 @@ import (
 	"context"
 	"crypto/tls"
 	"crypto/x509"
-	"errors"
 	"io/ioutil"
 	"log"
 	"os"
 
+	"github.com/alecthomas/kingpin"
```
[the remark this answers is not in the document]
Comment (ShaileshSurya) [reply cmd/contour/cli.go]: Done.

File: cmd/contour/cli.go [hunk 7]
```diff
@@ -49,17 +49,16 @@ func (c *Client) dial() *grpc.ClientConn {
 		}
 		// Load the client certificates from disk
 		certificate, err := tls.LoadX509KeyPair(c.ClientCert, c.ClientKey)
-		check(err)
-
+		kingpin.FatalIfError(err, "err: %v", err)
```
[the remark this answers is not in the document]
Comment (ShaileshSurya) [reply cmd/contour/cli.go]: Done.

**Comment by ShaileshSurya** [comment]
@jpeach I have made the requested changes. Can you please have a look.

File: cmd/contour/cli.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jpeach) [reply cmd/contour/cli.go]: It's better to provide a message string to give the context of the failure.  For example, this:
```
$ contour cli cds --cafile=/tmp/t --cert-file=/tmp/t --key-file=/tmp/t
contour: error: failed to load CA certificates: open /tmp/t: no such file or directory
```

is a better message than this:
```
$ contour cli cds --cafile=/tmp/t --cert-file=/tmp/t --key-file=/tmp/t
contour: error: open /tmp/t: no such file or directory
```

Could you please add a message string to each call to `kingpin.FatalIfError`?

File: cmd/contour/cli.go [hunk 8]
```diff
@@ -45,21 +44,20 @@ func (c *Client) dial() *grpc.ClientConn {
 	case c.CAFile != "" || c.ClientCert != "" || c.ClientKey != "":
 		// If one of the three TLS commands is not empty, they all must be not empty
 		if !(c.CAFile != "" && c.ClientCert != "" && c.ClientKey != "") {
-			log.Fatal("You must supply all three TLS parameters - --cafile, --cert-file, --key-file, or none of them.")
+			kingpin.Fatalf("You must supply all three TLS parameters - --cafile, --cert-file, --key-file, or none of them.")
```
Comment (jpeach) [inline cmd/contour/cli.go:47]: This is a pre-existing problem, but could you please remove the capital and fullstops from this message so it formats more consistently?

```suggestion
			kingpin.Fatalf("you must supply all three TLS parameters - --cafile, --cert-file, --key-file, or none of them")
```

File: cmd/contour/cli.go (hunk 1 above)
Comment (jpeach) [reply cmd/contour/cli.go:60]:
```suggestion
			kingpin.Fatalf("failed to append CA certs")
```

File: cmd/contour/cli.go (hunk 8 above)
Comment (ShaileshSurya) [reply cmd/contour/cli.go]: fixed this.

**Comment by ShaileshSurya** [comment]
@jpeach Please have a look, I have addressed the review comments.

File: cmd/contour/cli.go (hunk 5 above)
Comment (jpeach) [reply cmd/contour/cli.go:51]: This should be 
```
kingpin.FatalIfError(err, "failed to load certificate")
```

File: cmd/contour/cli.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jpeach) [reply cmd/contour/cli.go]: > Could you please add a message string to each call to `kingpin.FatalIfError`?

@ShaileshSurya ^^^

File: cmd/contour/cli.go (hunk 1 above)
Comment (jpeach) [reply cmd/contour/cli.go:60]: @ShaileshSurya ^^^

Comment (ShaileshSurya) [reply cmd/contour/cli.go:60]: Fixed this.

File: cmd/contour/cli.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShaileshSurya) [reply cmd/contour/cli.go]: I have made the required changes. We might want to check the wording for each and every error message once.

File: cmd/contour/cli.go (hunk 5 above)
Comment (ShaileshSurya) [reply cmd/contour/cli.go]: Made the changes to all FtalIfError. Added Error messages

**Review by jpeach — APPROVED** [review]
This looks good, thanks @ShaileshSurya

**Pull request merged**